src/utils/dockerDirManager.ts

```diff
@@ -0,0 +1,302 @@
+      Metadata:  { Description: 'Rancher Desktop moby context' },
+      Endpoints: {
+        docker: {
+          Host:          `unix://${ socketPath }`,
+          SkipTLSVerify: false,
+        },
+      } as Record<string, {Host: string, SkipTLSVerify: boolean, DefaultNamespace?: string}>,
+    };
+
+    if (kubernetesEndpoint) {
+      contextContents.Endpoints.kubernetes = {
+        Host:             kubernetesEndpoint,
+        SkipTLSVerify:    true,
+        DefaultNamespace: 'default',
+      };
+    }
+
+    console.debug(`Updating docker context: writing to ${ this.dockerContextPath }`, contextContents);
+
+    await fs.promises.mkdir(path.dirname(this.dockerContextPath), { recursive: true });
+    await fs.promises.writeFile(this.dockerContextPath, JSON.stringify(contextContents));
+  }
+
+  /**
+   * Clear the docker context; this is used for factory reset.
+   */
+  async clearDockerContext(): Promise<void> {
+    try {
+      await fs.promises.rm(path.dirname(this.dockerContextPath), { recursive: true, force: true });
+
+      const config = await this.readDockerConfig();
+
+      if (config?.currentContext !== this.contextName) {
+        return;
+      }
+      delete config.currentContext;
+      await this.writeDockerConfig(config);
+    } catch (ex) {
+      // Ignore the error; there really isn't much we can usefully do here.
+      console.debug(`Ignoring error when clearing docker context: ${ ex }`);
+    }
+  }
+
+  /**
+   * Ensures that the state of everything under the docker CLI config directory
+   * is correct.
+   * @param weOwnDefaultSocket Whether Rancher Desktop has control over the default socket.
+   * @param socketPath Path to the rancher-desktop specific docker socket.
+   * @param kubernetesEndpoint Path to rancher-desktop Kubernetes endpoint.
+   */
+  async ensureDockerConfig(weOwnDefaultSocket: boolean, socketPath?: string, kubernetesEndpoint?: string): Promise<void> {
+    // read current config
+    const currentConfig = await this.readDockerConfig();
+
+    console.log(`Read existing docker config: ${ JSON.stringify(currentConfig) }`);
+    const newConfig = JSON.parse(JSON.stringify(currentConfig));
+
+    // ensure docker context is set as we want
+    const platform = os.platform();
+
+    if ((platform === 'darwin' || platform === 'linux') && socketPath) {
+      await this.ensureDockerContext(socketPath, kubernetesEndpoint);
+    }
+    newConfig.currentContext = await this.getDesiredDockerContext(weOwnDefaultSocket, currentConfig.currentContext);
+
+    // ensure we are using a valid credential helper
+    if (!newConfig.credsStore) {
+      newConfig.credsStore = this.getDefaultDockerCredsStore();
+    } else if (newConfig.credsStore === 'desktop' && !(await this.credHelperWorking(newConfig.credsStore))) {
+      newConfig.credsStore = this.getDefaultDockerCredsStore();
+    }
+
+    // write config if modified
+    if (JSON.stringify(newConfig) !== JSON.stringify(currentConfig)) {
```

Comment:
I believe JSON stringification is not guaranteed to be stable; erroring on writing more is probably fine though.

Reply:
So this is fine? I think we could use lodash for this, but I wanted to avoid dependencies if we didn't need them.

Reply:
Yeah, I think it's okay for now; we can pull in things if we find that it's inadequate later.